chan.go

```diff
@@ -83,26 +109,29 @@ func (r *chanResponse) Next() (interface{}, error) {
 	}
 
 	select {
+	case <-r.closed:
+		return nil, r.err
 	case v, ok := <-r.ch:
 		if !ok {
+			r.ch = nil
```

Comment:
Are you sure this isn't racy? Also, this'll cause this method to hang forever if called again.